### DISCUSSION TITLE
How to check if a read from an input stream succeeds?

### DISCUSSION THREAD

**Discussion by DUOLabs333** [capnproto/capnproto#1915] [Q&A] [answered]
I have a system that incrementally reads from an  asio::bufstream into a STL ifstream. I want to stop copying once Capn successfully reads the data in the ifstream, something like the `accept` function in nlohmann's json library.

The best I can come up with is to compute the message size in bytes, write it to the bufstream as fr"{size}\n", read_until('\n'), read the number, then transfer_exactly that many bytes. I'm not sure this is efficient though.

**Comment by kentonv** [comment]
A size prefix would be somewhat redundant as Cap'n Proto messages are self-delimiting. However, capnp's encoding isn't quite so simple as a size prefix, instead you have to read the whole "segment table" before you can compute the full message size. This is described at:

https://capnproto.org/encoding.html#serialization-over-a-stream

In short, reading the first 4 bytes will tell you the segment table size, and then after reading the whole segment table you can compute the full message size.

You might find `expectedSizeInWordsFromPrefix()` useful here:

https://github.com/capnproto/capnproto/blob/master/c%2B%2B/src/capnp/serialize.h#L108

If this all seems too complicated, then writing your own size prefix may be reasonable. But, I'd recommend writing it as a fixed-width binary value rather than text.

**Reply by DUOLabs333** [reply]
The problem I have with using the raw bytes from a `uint64_t`, is endianness --- interpreting text is slower, but it will work. Or did you mean compute the 1's and 0's for the size (let's say 64 bit), send that array, then read the 64 bits back?

Yeah, I did look at `expectedSizeInWordsFromPrefix()` but it's possible that it's an underestimate, so I didn't want to take that risk. serialize.h is how I learned about ` computeSerializedSizeInWords` --- though how big is a word?

**Reply by kentonv** [reply, marked as answer]
Typically you would read an integer in an endianness-independent way like so:

    unsigned char buf[4];
    read(buf, 4);
    uint32_t value = buf[0] | (buf[1] << 8) | (buf[2] << 16) | (buf[3] << 24);

Note that this code correctly reads a little-endian value from the wire regardless of whether the host endianness is big or little.

(That said, all modern architectures are little-endian, so IMO it would also be perfectly fine to stop worrying about byte order and just write integers directly to the wire.)

> though how big is a word?

8 bytes.

**Reply by DUOLabs333** [reply]
Thanks --- just one more question: this means that  messageToFlatArray().asBytes().size()==8*computeSerializedSizeInWords(), right?

**Reply by kentonv** [reply]
Yes. You can also use `sizeof(capnp::word)` if you don't like the magic number 8.

**Reply by DUOLabs333** [reply]
Awesome --- thanks!
